File: Benchmarks/OOP/OOP.py
class Vector3:
    __slots__ = ['x', 'y', 'z']
    def __init__(self, x: float, y: float, z: float):
        self.x = x
        self.y = y
        self.z = z

    def magnitude_squared(self) -> float:
        return self.x * self.x + self.y * self.y + self.z * self.z

    def add(self, other: 'Vector3') -> 'Vector3':
        return Vector3(self.x + other.x, self.y + other.y, self.z + other.z)


class Loggable:
    def __init__(self):
        self.log_count = 0
    def log_message(self, message: str) -> None:
        self.log_count += 1
    def get_log_count(self) -> int:
        return self.log_count


class Shape(Loggable):
    def __init__(self, x: float, y: float):
        super().__init__()
        self.x = x
        self.y = y


class Circle(Shape):
    def __init__(self, x: float, y: float, radius: float):
        super().__init__(x, y)
        self.radius = radius

    def area(self) -> float:
        return 3.141592653589793 * self.radius * self.radius

    def perimeter(self) -> float:
        return 2.0 * 3.141592653589793 * self.radius


class Rectangle(Shape):
    def __init__(self, x: float, y: float, width: float, height: float):
        super().__init__(x, y)
        self.width = width
        self.height = height

    def area(self) -> float:
        return self.width * self.height

    def perimeter(self) -> float:
        return 2.0 * (self.width + self.height)


class FileHandle:
    def __init__(self, path: str):
        self.path = path
        self.is_open = True

    def read(self) -> str:
        return "Content" if self.is_open else "Closed"

    def close(self) -> None:
        self.is_open = False

    def finalize(self) -> None:
        self.close()


v1 = Vector3(1.0, 2.0, 3.0)
v2 = Vector3(4.0, 5.0, 6.0)
sum_struct = 0.0
i = 0
while i < 10000000:
    res = v1.add(v2)
    sum_struct += res.magnitude_squared()
    i += 1
print(f"Struct: {sum_struct}")

sum_area = 0.0
j = 0
while j < 5000000:
    if j % 2 == 0:
        shape = Circle(0.0, 0.0, 5.0)
    else:
        shape = Rectangle(0.0, 0.0, 10.0, 20.0)
    sum_area += shape.area()
    j += 1
print(f"Polymorphism: {sum_area}")

k = 0
while k < 3000000:
    file = FileHandle("/tmp/test.txt")
    file.read()
    file.finalize()
    k += 1
print("RAII Completed.")

File: Benchmarks/OOP/OOP.js
class Vector3 {
    constructor(x, y, z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }
    magnitude_squared() {
        return this.x * this.x + this.y * this.y + this.z * this.z;
    }
    add(other) {
        return new Vector3(this.x + other.x, this.y + other.y, this.z + other.z);
    }
}


class Loggable {
    constructor() {
        this.log_count = 0;
    }
    log_message(message) {
        this.log_count++;
    }
    get_log_count() {
        return this.log_count;
    }
}


class Shape extends Loggable {
    constructor(x, y) {
        super();
        this.x = x;
        this.y = y;
    }
}


class Circle extends Shape {
    constructor(x, y, radius) {
        super(x, y);
        this.radius = radius;
    }
    area() {
        return 3.141592653589793 * this.radius * this.radius;
    }
    perimeter() {
        return 2.0 * 3.141592653589793 * this.radius;
    }
}


class Rectangle extends Shape {
    constructor(x, y, width, height) {
        super(x, y);
        this.width = width;
        this.height = height;
    }
    area() {
        return this.width * this.height;
    }
    perimeter() {
        return 2.0 * (this.width + this.height);
    }
}


class FileHandle {
    constructor(path) {
        this.path = path;
        this.is_open = true;
    }
    read() {
        return this.is_open ? "Content" : "Closed";
    }
    close() {
        this.is_open = false;
    }
    finalize() {
        this.close();
    }
}


let v1 = new Vector3(1.0, 2.0, 3.0);
let v2 = new Vector3(4.0, 5.0, 6.0);
let sum_struct = 0.0;
let i = 0;
while (i < 10_000_000) {
    let res = v1.add(v2);
    sum_struct += res.magnitude_squared();
    i++;
}
console.log("Struct: " + sum_struct);

let sum_area = 0.0;
let j = 0;
while (j < 5_000_000) {
    let shape;
    if (j % 2 === 0) {
        shape = new Circle(0.0, 0.0, 5.0);
    } else {
        shape = new Rectangle(0.0, 0.0, 10.0, 20.0);
    }
    sum_area += shape.area();
    j++;
}
console.log("Polymorphism: " + sum_area);

let k = 0;
while (k < 3_000_000) {
    let file = new FileHandle("/tmp/test.txt");
    file.read();
    file.finalize();
    k++;
}
console.log("RAII Completed.");
